doc/otp-sct.1
...\" $Header: /usr/src/docbook-to-man/cmd/RCS/docbook-to-man.sh,v 1.3 1996/06/17 03:36:49 fld Exp $
...\"
...\"	transcript compatibility for postscript use.
...\"
...\"	synopsis:  .P! <file.ps>
...\"
.de P!
\\&.
.fl			\" force out current output buffer
\\!%PB
\\!/showpage{}def
...\" the following is from Ken Flowers -- it prevents dictionary overflows
\\!/tempdict 200 dict def tempdict begin
.fl			\" prolog
.sy cat \\$1\" bring in postscript file
...\" the following line matches the tempdict above
\\!end % tempdict %
\\!PE
\\!.
.sp \\$2u	\" move below the image
..
.de pF
.ie     \\*(f1 .ds f1 \\n(.f
.el .ie \\*(f2 .ds f2 \\n(.f
.el .ie \\*(f3 .ds f3 \\n(.f
.el .ie \\*(f4 .ds f4 \\n(.f
.el .tm ? font overflow
.ft \\$1
..
.de fP
.ie     !\\*(f4 \{\
.	ft \\*(f4
.	ds f4\"
'	br \}
.el .ie !\\*(f3 \{\
.	ft \\*(f3
.	ds f3\"
'	br \}
.el .ie !\\*(f2 \{\
.	ft \\*(f2
.	ds f2\"
'	br \}
.el .ie !\\*(f1 \{\
.	ft \\*(f1
.	ds f1\"
'	br \}
.el .tm ? font underflow
..
.ds f1\"
.ds f2\"
.ds f3\"
.ds f4\"
.ta 8n 16n 24n 32n 40n 48n 56n 64n 72n 
.TH "\fBotp-sct\fP" "1"
.SH "NAME"
\fBotp-sct\fP \(em Smart Card Terminal for One Time Password package\&.
.SH "SYNOPSIS"
.PP
\fBotp-sct\fP [-1hlpv?]  [-c\fI count\fP]  [-d\fI debug_level\fP]  [-i\fI index\fP]  [-r\fI reader\fP]  [-v\fI firmware_api_version\fP] 
.SH "DESCRIPTION"
.PP
The \fBotp-sct\fP command is a user interface to generating
One Time Passwords with a Smart Card loaded with OTP software\&.
.SH "OPTIONS"
.IP "-c, --sc-count=\fI count\fP" 10
Configure the optional count parameter to sync a Smart Card to a challenge\&.
A count value of 0 indicates the HOTP value is to be calculated with the
current stored count\&.
.IP "-d, --debug=\fI debug_level\fP" 10
Set debug level\&.
.IP "-h, --help" 10
Help\&.
.IP "-i, --sc-index=\fI index\fP" 10
Set the 8 bit index\&.  The Smart Card contains numerically indexed records
for each host system\&.  Use the -l option to list hostnames associated with
the index\&.  The default index is 0\&.
.IP "-l, --list-readers" 10
List SC Readers
.IP "-L, --sc-list-hostnames" 10
List host systems configured on the Smart Card\&.  The index is used with the -i option\&.
.IP "-p, --sc-reset-pin" 10
Set new PIN\&.
.IP "-r, --reader=\fI reader\fP" 10
Set Smart Card reader\&.  Use -l to list available readers\&.  A reader
is defined as class:reader:[option]\&.  PCSC and embedded
are the two available classes\&.  The embedded class contains the acr30s driver
which is specified as embedded:acr30s:[serial_port]\&.
If pcscd is running the first PC/SC reader will be the default followed by
the embedded acr30s driver\&.  Use PCSC: for the first available PC/SC
reader\&.  Use embedded:acr30s:/dev/cuaU0 for the embedded acr30s driver
with serial port /dev/cuaU0\&.
.IP "-v, --sc-version=\fI card_api_version\fP" 10
Set the Smart Card API version\&.  The binary API between the terminal
and Smart Card changed between version 2 and 3\&.  See command mode notes
above\&.  The default version is 3\&.  Configuring version 2 will allow
maintenance of Smart Card with version 2 firmware\&.
.IP "-V, --sc-list-version" 10
List the Smart Card firmware version\&.
.IP "-1, --sc-get-hostp-v1" 10
Use the version 1 GetHOTP command instead of the default GetHOTPHostCount32\&.
The latter is not available on firmware revision 1\&.  GetHOTP may be conditionally
compiled out of newer firmware\&.
.IP "--version" 10
Display software version\&.
.SH "EXAMPLES"
.PP
Generate a HOTP for the first system on the first PCSC reader found\&.
.PP
.nf
\fBotp-sct -r PCSC:\fP 
\f(CWEnter PIN: 12345
HOTP: dev1\&.eng 2A5AB4B78D\fP
.fi
.PP
List systems configured on Smart Card in default reader\&.  Generate HOTP
for dev3\&.eng with count 32\&.
.PP
.nf
\fBotp-sct -L\fP 
\f(CWEnter PIN: 12345
00:dev1\&.eng
01:dev2\&.eng
02:dev3\&.eng
03:vpn1\&.eng\fP 
\fBotp-sct -i 2 -c 32\fP 
\f(CWEnter PIN: 12345
HOTP: dev3\&.eng 2A9BE3F142\fP
.fi
.PP
Note that hosts with the READERKEY flag set will not be displayed\&.
\fBotp-sct\fP can not set the reader key\&.
.SH "AUTHOR"
.PP
Mark Fullmer maf@splintered\&.net
.SH "SEE ALSO"
.PP
\fBotp-sca\fP(1)
\fBotp-control\fP(1)
\fBpam_otp\fP(1)
\fBhtsoft-downloader\fP(1)
\fBotp-ov-plugin\fP(1)
\fBbcload\fP(1)
\fBurd\fP(1)
spyrus-par2(7)
...\" created by instant / docbook-to-man, Thu 12 Dec 2013, 10:40


doc/CHANGES
# $Id:$

htsoft-downloader: Makefile install target fix - eshicks

pam_otp: allown_unknown typo - eshicks

otplib: more sanity checking on otp_user_add - eshicks

otp-tokend: smtp support, fork() on message

urd/rc.d/urd: updated for new flags

otp-control: mode generate will honor -c

send-tokendd: reference implementation of OOB token.  Sends token to web
page via ssl.  Used internally for SMS gateway.

send-token feature.  Send to token to user via flexible OOB method.

HOTPC.BAS rev 9.  Check PIN, readerKey

otp-sca support SC_SETREADERKEY_CAP

pam_otp.so: require_db_entry is now defaulted to enabled.  A new option
allow_unknown_user which is mutually exclusive to require_db_entry is
accepted for consistency with urd and openvpn.  kyriacou@osc.edu

otpdb2sc missing quote.  kyriacou@osc.edu

openvpn plugin accepts otp-allow-unknown-user

otp-control will by default create a new user with count=1.  BasicCard
code treats count=0 as a special case when a challenge (count) is supplied.
0 is ignore challenge and use EEPROM stored count.

getopt_long() support.  All binaries support --version.

honor OTP_FLAGS_DISPNT in urd.  Use -c to force display

urd will omit password when packed decode debug is enabled

fix state hex encoding in urd when state > 127

fix soft reload option in urd

add PAM password authentication to urd

str.c: tcsetattr() missing in str_input()

otp-control -m list-sc will automatically set the correct FMT flags

document behavior of count=0 in otp-sct, otp-sca, and spyrus-reader

add urd/rc.d script to dist

add rfc.card to dist

spyrus1.4.hex installed instead of 1.3.  Also in QUICKSTART

### ootp-1.03.tar.gz

add CHANGES and TODO to dist

balance card enabled by default with HOTPC

finish dec31.x and DHEX40 otp-sca, otp-sct

add -i option to htsoft-downloader to ignore last WOK timeout

reorganize BasicCard code to minimize RAM and Flash resources.  Fix
PRDisplay (balance reader) syntax error intruduced with BalanceCardIndex

### snap-12-16-2009

add dhex40 format

add set-flags

remove activate, disable, deactivate, flags-dspcnt, flags-no-dspcnt
from otp-control.  Add more generic set-flags, set-format, set-status,
and set-type.

add dec31.6 dec31.7 dec31.8 dec31.9 dec31.10 formats.

add runtime window option to otp-openvpn, urd, and pam_otp.

scr: default to embedded reader if pcscd is not running.  PCSC: defaults
 to first PCSC reader.

urd: -O not positionally dependent on -o. Can disable OTP and specify otpdb.

add FreeBSD rc.d script for urd

add Zeitcontrol balance reader datasheet to docs

### ootp-1.02.tar.gz

otpdb2sc: do not right 0 pad index to otp-control

pam_otp.so: compile PIC

### ootp-1.01.tar.gz

release: build man and html pages at release time

sccmd.c: SetAdminKey fw version check reversed

urd: build default HOTP enabled

### ootp-1.00.tar.gz


otp-sct/otp-sct.c
/*
 * Copyright (c) 2005 Mark Fullmer
 * Copyright (c) 2009 Mark Fullmer and the Ohio State University
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *      $Id: otp-sct.c 131 2010-06-15 14:26:23Z maf $
 */

#include <sys/cdefs.h>
#include <sys/param.h>
#include <sys/types.h>
#include <sys/errno.h>
#include <getopt.h>
#include <inttypes.h>
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <ctype.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <termios.h>
#include "scr.h"
#include "sccmd.h"
#include "str.h"
#include "xerr.h"
#include "otpsc.h"
#include "otplib.h"

#if defined(__FreeBSD__)
#include <sys/endian.h>
#endif

#if defined(__DARWIN_UNIX03)
#include <sys/_endian.h>
#endif

#define OTP_ERROR          -1      /* library function failure */
#define OTP_SUCCESS        0       /* library function success */
#define OTP_FAIL           1       /* library function failure */

#define BZS(A) bzero(A, sizeof A);

#define SWAP32(x) x = \
         ((((x)&0xff)<<24) |\
         (((x)&0xff00)<<8) |\
         (((x)&0xff0000)>>8) |\
         (((x)>>24)&0xff));

#define SWAP16(x) x = \
    ( (((x)&0xff)<<8) | (((x)&0xff00)>>8) );

static int debug;

void help(void);


int main(int argc, char **argv)
{
  extern char *ootp_version;
  struct scr_ctx *scrctx;
  int i, j, k, r, sc_idx_set, sc_idx_tmp, j_start, j_end, opt_version;
  int reset_pin, list_readers, list_version, get_hotp_version, list_hostnames;
  uint32_t tmp_count, tmp32u;
  uint64_t tmp64u;
  char sc_hostname[SC_HOSTNAME_LEN+1], sc_pin[SC_PIN_LEN+1];
  char sc_newpin[SC_PIN_LEN+1], sc_newpin2[SC_PIN_LEN+1];
  char fmt_buf[133], fmt_buf2[133];
  u_char sc_hotp[SC_HOTP_LEN+1], sc_idx[SC_INDEX_LEN+1];
  u_char sc_version[SC_VERSION_LEN+1], sc_count32[SC_COUNT32_LEN+1];
  u_char sc_fv;
  char *reader, *endptr, *err_msg;

  struct option longopts[] = {
    { "sc-get-hotp-v1",             0, (void*)0L, '1'},
    { "sc-count",                   1, (void*)0L, 'c'},
    { "debug",                      1, (void*)0L, 'd'},
    { "help",                       0, (void*)0L, 'h'},
    { "help",                       0, (void*)0L, '?'},
    { "sc-index",                   1, (void*)0L, 'i'},
    { "list-readers",               0, (void*)0L, 'l'},
    { "sc-list-hostnames",          0, (void*)0L, 'L'},
    { "sc-reset-pin",               0, (void*)0L, 'p'},
    { "reader",                     1, (void*)0L, 'r'},
    { "sc-version",                 1, (void*)0L, 'v'},
    { "sc-list-version",            0, (void*)0L, 'V'},
    { "version",                    0, &opt_version, 1},
    { 0, 0, 0, 0},
  };

  /* init xerr */
  xerr_setid(argv[0]);

  opt_version = 0;
  sc_fv = 5;
  debug = 0;
  sc_idx_set = 0;
  reset_pin = 0; /* no */
  tmp_count = 0;
  reader = (char*)0L;
  list_readers = 0; /* no */
  list_version = 0; /* no */
  list_hostnames = 0; /* no */
  scrctx = (struct scr_ctx*)0L;
  get_hotp_version = 3;

  BZS(sc_hotp);
  BZS(sc_idx);
  BZS(sc_version);
  BZS(sc_count32);
  BZS(sc_hostname);
  BZS(sc_pin);
  BZS(sc_newpin);

  bcopy(SC_PIN_DEFAULT, sc_pin, SC_PIN_LEN);
  bcopy(SC_PIN_DEFAULT, sc_newpin, SC_PIN_LEN);

  while ((i = getopt_long(argc, argv, "1c:d:hi:lLpr:v:V?", longopts,
    (int*)0L)) != -1) {

    switch (i) {

      case '1':
        get_hotp_version = 1;
        break;

      case 'c':
        tmp64u = strtoull(optarg, &endptr, 0);
        if (*endptr)
          xerr_errx(1, "strtoull(%s): failed at %c.", optarg, *endptr);
        if (tmp64u > SC_COUNT32_MAX)
          xerr_errx(1, "count > SC_COUNT32_MAX.");
        tmp_count = tmp64u;
#if BYTE_ORDER == LITTLE_ENDIAN
        SWAP32(tmp_count);
#endif /* LITTLE_ENDIAN */
        bcopy(&tmp_count, sc_count32, 4);
        tmp_count = 1; /* signal was set */
        break;

      case 'd':
        debug = strtoul(optarg, &endptr, 0);
        if (*endptr)
          xerr_errx(1, "strtoul(%s): failed at %c.", optarg, *endptr);
        break;

      case 'h':
      case '?':
        help();
        exit(0);
        break; /* notreached */

      case 'i':
        sc_idx_tmp = strtoul(optarg, &endptr, 0);
        if (*endptr)
          xerr_errx(1, "strtoul(%s): failed at %c.", optarg, *endptr);
        if (sc_idx_tmp > SC_INDEX_MAX) {
          xerr_errx(1, "Index out of range 0..%d", SC_INDEX_MAX);
        } else {
          sc_idx[0] = sc_idx_tmp;
          sc_idx_set = 1;
        }
        break;

      case 'l':
        list_readers = 1;
        break;

      case 'L':
        list_hostnames = 1;
        break;

      case 'p':
        reset_pin = 1;
        break;

      case 'r':
        reader = optarg;
        break;

      case 'v':
        sc_fv = strtoul(optarg, &endptr, 0);
        if (*endptr)
          xerr_errx(1, "strtoul(%s): failed at %c.", optarg, *endptr);
        get_hotp_version = 1;
        break;

      case 'V':
        list_version = 1;
        break;

     case 0:
        if (opt_version) {
          printf("%s\n", ootp_version);
          exit(0);
        }

     default:
        xerr_errx(1, "getopt_long(): fatal.");

    } /* switch */

  } /* while getopt_long() */

  /* get pin */
  if (!list_readers && !list_version) {

    while (1) {

      if (str_input("Enter PIN: ", sc_pin, SC_PIN_LEN+1,
        STR_FLAGS_ECHO_OFF) < 0)
        xerr_errx(1, "str_input(%d): failed.", SC_PIN_LEN);

      if (strlen((char*)sc_pin) != SC_PIN_LEN) {

        xerr_warnx("PIN Must be %d characters", SC_PIN_LEN);
        continue;

      } else {

        break;

      } /* SC_PIN_LEN */

    } /* valid pin len */

  } /* need PIN */

  if (!(scrctx = scr_ctx_new(SCR_READER_EMBEDDED_ACR30S|SCR_READER_PCSC,
    debug))) {
    xerr_errx(1, "scr_ctx_new(): failed");
  }

  if (list_readers) {

    for (i = 0; i < scrctx->num_readers; ++i)
      printf("%s\n", scrctx->readers[i]);

    goto main_out;

  }

  /* get new pin? */
  if (reset_pin) {

    while (1) {

      if (str_input("New PIN: ", sc_newpin, SC_PIN_LEN+1,
        STR_FLAGS_ECHO_OFF) < 0)
        xerr_errx(1, "str_input(%d): failed.", SC_PIN_LEN);

      if (strlen((char*)sc_newpin) != SC_PIN_LEN) {
        xerr_warnx("PIN Must be %d characters.", SC_PIN_LEN);
        continue;
      }

      if (str_input("New PIN (again): ", sc_newpin2, SC_PIN_LEN+1,
        STR_FLAGS_ECHO_OFF) < 0)
        xerr_errx(1, "str_input(%d): failed.", SC_PIN_LEN);

      /* identical? then done */
      if (!bcmp(sc_newpin,sc_newpin2, SC_PIN_LEN))
        break;

      printf("New PIN did not match, try again.\n");

    } /* new pin confirm */

  } /* reset_pin */

  if (scr_ctx_connect(scrctx, reader) < 0) {
    xerr_errx(1, "scr_ctx_connect(): failed");
  }

/****************/

  if (list_version) {

    if ((r = sccmd_GetVersion(scrctx, sc_fv, sc_version)) < 0) {
      xerr_errx(1, "sccmd_GetVersion(): failed.");
    }

    if (r == 0) {
      str_hex_dump(fmt_buf, sc_version, 1);
      printf("Version: 0x%s\n", fmt_buf);
    } else if (r == 1) {
      printf("Version: fail\n");
    } else {
      xerr_errx(1, "sccmd_GetVersion(): fatal.");
    }

    goto main_out;

  } /* list_version */

/****************/

  if (reset_pin) {

    if ((r = sccmd_SetPIN(scrctx, sc_fv, sc_pin, sc_newpin)) < 0) {
      xerr_errx(1, "sccmd_SetPIN(): failed.");
    }

    if (r == 0)
      printf("SetPIN Good.\n");
    else if (r == 1)
      printf("SetPIN Bad.\n");
    else
      xerr_errx(1, "sccmd_SetPIN(): fatal.");

    goto main_out;

  } /* reset_pin */

  if (list_hostnames) {

    /* if no index, iterate until empty hostname */
    if (sc_idx_set) {
      j_start = sc_idx[0];
      j_end = sc_idx[0];
    } else {
      j_start = 0, j_end = SC_INDEX_MAX;
    }

    for (j = j_start; j <= j_end; ++j) {

      sc_idx[0] = j;
  
      if ((r = sccmd_GetHostName(scrctx, sc_fv, sc_idx, sc_pin,
        sc_hostname)) < 0)
        xerr_errx(1, "sccmd_GetHostName(): failed.");
  
      if (r == 0) {
  
        /* empty hostname is end of data */
        if (sc_hostname[0] == 0)
          break;
 
        str_hex_dump(fmt_buf, sc_idx, 1); 
        i = (1<<1); fmt_buf[i] = ':'; i += 1;

        /* required reader key then skip */
        if (sc_hostname[HOSTNAME_POS_READERKEY] & HOSTNAME_FLAG_MASK) {

          continue;

        } /* READERKEY flag */


        for (k = 0; k < SC_HOSTNAME_LEN; ++k) {

          fmt_buf[i++] = sc_hostname[k] & ~HOSTNAME_FLAG_MASK;

          if (sc_hostname[k] == 0)
            break;

        } /* hostname chars */
  
        printf("%s\n", fmt_buf);
  
      } else if (r == 1) {
  
        printf("GetHostName reject\n");
        break;
  
      } else {
  
        xerr_errx(1, "sccmd_GetHostName(): fatal.");
  
      }
  
    } /* for host */

    goto main_out;

  } /* list_hostnames */

/****************/  

  if (get_hotp_version == 1) {

    if ((r = sccmd_GetHOTP(scrctx, sc_fv, sc_idx, sc_pin, sc_hotp)) < 0)
      xerr_errx(1, "sccmd_GetHOTP(): failed.");

      err_msg = "sccmd_GetHOTP(): fatal.";

  } /* get_hotp_version == 1 */

  if (get_hotp_version == 3) {

    if ((r = sccmd_GetHOTPHostCount32(scrctx, sc_fv, sc_idx, sc_pin,
      sc_count32, sc_hotp, sc_hostname)) < 0)
      xerr_errx(1, "sccmd_GetHOTPHostCount32(): failed.");
    
    err_msg = "sccmd_GetHOTPHostCount32(): fatal.";

  } /* get_hotp_version == 3 */

  /* successful SC transaction? */
  if (r == 0) {

    if (sc_hostname[HOSTNAME_POS_FMT] & HOSTNAME_FLAG_MASK) {

      tmp32u = (sc_hotp[0] << 24) | (sc_hotp[1] << 16) |
               (sc_hotp[2] << 8) | sc_hotp[3];

      str_uint32toa(fmt_buf, tmp32u);

    } else {

      str_hex_dump(fmt_buf, sc_hotp, 5);

    }

    for (i = 0, j = 0; i < SC_HOSTNAME_LEN; ++i) {

      /* clear high bit for display */
      sc_hostname[i] &= ~HOSTNAME_FLAG_MASK;

    }


    if (get_hotp_version == 3) {
      str_ftoc(fmt_buf2, sc_hostname, SC_HOSTNAME_LEN);
      printf("HOTP: %s %s\n", fmt_buf2, fmt_buf);
    }

    if (get_hotp_version == 1) {
      printf("HOTP: %s\n", fmt_buf);
    }
      
  } else if (r == 1) {
    printf("HOTP: rejected\n");
  } else {
    xerr_errx(1, err_msg);
  }

main_out:

  scr_ctx_free(scrctx);

  exit (0);

} /* main */

void help(void)
{
  fprintf(stderr, "otp-sct [-1hlpv?] [-c count] [-d debug_level] [-i index]\n");
  fprintf(stderr, "        [-r reader] [-v card_api_version]\n");
  fprintf(stderr, "        -1 : Use version 1 firmware GetHOTP\n");
  fprintf(stderr, "        -h : help\n");
  fprintf(stderr, "        -l : list SC readers\n");
  fprintf(stderr, "        -L : list hostnames\n");
  fprintf(stderr, "        -p : reset PIN\n");
  fprintf(stderr, "        -V : list SC firmware version\n");
} /* help */

